### ISSUE TITLE
Massive amount of linking errors when importing sciplot in multiple files.

### ISSUE THREAD

**Issue by CaeruleusAqua** [sciplot/sciplot#32]
Hello together,

has anyone tried to use sciplot in a larger c++ project?

Sciplot claims to be a header only library without being implemented as such.
In the headers are tons of (non template ) function definitions outside the class definition. This results in a lot of double definition errors when linking.

Also, in util.hpp all non-template functions should be declared inline.

I can submit a patch for this if you can confirm the error.

Greetings,
Julian

**Comment by allanleal** [comment]
Hi Julian. Repeating the message here for convenience. 

Please check version `v0.2` here:
https://github.com/sciplot/sciplot/tree/v0.2

This is an almost completely rewritten version of sciplot. The issue reported here is not present there (all inline methods properly marked `inline`). 

I'll put the CI to work correctly for v0.2 (the CI is broken here because I just to the GH actions from another project I have) and then merge. And also produce an updated website on https://sciplot.github.io.

**Comment by CaeruleusAqua** [comment]
Hi Allanleal,

I checkt version v0.2. And get the same errors.

**Comment by allanleal** [comment]
OK, I'll fix this.

**Comment by allanleal** [comment]
OK, I've added a few more missing inline specifiers in the v0.2 branch. I've create this branch

https://github.com/sciplot/sciplot/compare/v0.2-errors-fixed

that contains your `larger-cmake-project`. No linking errors anymore.

Another thing. I'm wondering if having `sciplot` as a compiled library would be more preferable. I'm concerned about compilation times growing with time for dependent projects. @HorstBaerbel , I would like your opinion as well. Also, in this v0.2 version, I've adopted the use of camel case as you suggested before.

**Comment by HorstBaerbel** [comment]
I guess I was responsible for lots of the missing inlines ;), so thanks for fixing that! Nice work merging plot and multiplot, though I somehow prefer "plot" to "figure". All-in-all it definitely looks better!
I still think it might be a good idea to add a "canvas" class as proposed in #14 . Maybe @CaeruleusAqua / Julian has an developer + user opinion on that too...

**Comment by HorstBaerbel** [comment]
About the library part: Header-only libraries have a certain appeal to me, because of no linking, or architecture problems, deployment / packaging is easy etc. I don't see the library becoming too big to be header-only. There's much bigger ones.
Compile times will increase though, indeed, and if you're not careful you can get linker errors etc.
So: I'm definitely undecided ;)

**Comment by CaeruleusAqua** [comment]
> I guess I was responsible for lots of the missing inlines ;), so thanks for fixing that! Nice work merging plot and multiplot, though I somehow prefer "plot" to "figure". All-in-all it definitely looks better!
> I still think it might be a good idea to add a "canvas" class as proposed in #14 . Maybe @CaeruleusAqua / Julian has an developer + user opinion on that too...

I don't want to answer the fundamental question of whether we create a "canvas" class. Looking at the new Figure.hpp, it's probably a good idea to split it up.
At the moment it looks a little bit like god class

**Comment by allanleal** [comment]
> So: I'm definitely undecided ;)

Better to keep header-only for a while, until this proves inadequate. 

> At the moment it looks a little bit like god class

I also agree it's quite long. This is due to the fact that now we have one method for every plot function. Some plots have one or more overloads (`drawBoxes(x, y)`, `drawBoxes(x, y, widths)`). 

Moreover, each such method also accepts data in file form (e.g., `drawBoxes("data.dat", 1, 5, 2)`, where 1, 2, 3 are column indices).

I'm open to suggestions on how it could be refactored.

> though I somehow prefer "plot" to "figure"

@CaeruleusAqua  are you also in favor of `Plot` over `Figure`, like @HorstBaerbel ? Options below:

**Option 1:** `Figure`, `Figures`.
**Option 2:** `Plot`, `Multiplot`.
**Option 3:** `Plot`, `Plots`.

**Comment by CaeruleusAqua** [comment]
Here is my informed opinion on header only libraries:

Header only libraries only make sense from my point of view if I have a very small library, preferably consisting of only one file. 

I don't see any problems with deployment / packaging with a proper build system like cmake. A few years ago this looked certainly different.

For packaging and deployment you can simply use cpack. And this also on all systems (Linux / Windows / Mac). 

As long as the library is used in a cmake project, there is nothing more to do than:

```
find_package(sciplot)
target_link_libraries(your_taget sciplot::sciplot)
```

It makes no difference if it is header only or not.

The "installation" is also trivial. Just build the library from the repo and compile it. Done! Cmake already exports all information about the library in the "cmake" step. From the developer perspective this has the advantage that all projects automatically use the current developer version. You could also export the developer version under a different name (e.g. sciplot-dev), if this behavior is not wanted.

**Comment by CaeruleusAqua** [comment]
> > So: I'm definitely undecided ;)
> 
> Better to keep header-only for a while, until this proves inadequate.
> 
> > At the moment it looks a little bit like god class
> 
> I also agree it's quite long. This is due to the fact that now we have one method for every plot function. Some plots have one or more overloads (`drawBoxes(x, y)`, `drawBoxes(x, y, widths)`).
> 
> Moreover, each such method also accepts data in file form (e.g., `drawBoxes("data.dat", 1, 5, 2)`, where 1, 2, 3 are column indices).
> 
> I'm open to suggestions on how it could be refactored.
> 
> > though I somehow prefer "plot" to "figure"
> 
> @CaeruleusAqua are you also in favor of `Plot` over `Figure`, like @HorstBaerbel ? Options below:
> 
> **Option 1:** `Figure`, `Figures`.
> **Option 2:** `Plot`, `Multiplot`.
> **Option 3:** `Plot`, `Plots`.

I would go for `Plot`, Multiplot`,  `Figure`, `Figures`. looks to similar

**Comment by allanleal** [comment]
> Header only libraries only make sense from my point of view if I have a very small library, preferably consisting of only one file.

Also important to mention when the library template-only. This is the case in this other lib I develop [autodiff.github.io](https://autodiff.github.io).

> I don't see any problems with deployment / packaging with a proper build system like cmake. A few years ago this looked certainly different.

I agree here as well. I've being using conda for managing my dependencies (and I know you, @CaeruleusAqua uses vcpkg for achieving the same result). These are great solutions.

**Comment by CaeruleusAqua** [comment]
In general, I would probably prefer something like a Figure and Plot class, where a Figure can have multiple Plots. Multiplots would then be generated automatically by adding new plots to the figure.

Which is indeed similar to what @HorstBaerbel suggested.

**Comment by allanleal** [comment]
> In general, I would probably prefer something like a Figure and Plot class, where a Figure can have multiple Plots. Multiplots would then be generated automatically by adding new plots to the figure.
> Which is indeed similar to what @HorstBaerbel suggested.

OK, this makes sense.

**Comment by CaeruleusAqua** [comment]
Looks like a fancy solution. I'll wait for the final solution on Monday when I continue working on my stuff ;-)

**Comment by allanleal** [comment]
This has been fixed in v0.2.
